### PR TITLE
Move sigterm handling to task level

### PR DESCRIPTION
**Pull request by mkosiarc**
CLOUDBLD-10115

Signed-off-by: mkosiarc <mkosiarc@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] Python type annotations added to new code
- n/a JSON/YAML configuration changes are updated in the relevant schema
- n/a Changes to metadata also update the documentation for the metadata
- n/a Pull request has a link to an osbs-docs PR for user documentation updates
- n/a New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/tasks/common.py [hunk 1]
```diff
@@ -90,6 +93,8 @@ class Task(abc.ABC, Generic[ParamsT]):
     def __init__(self, params: ParamsT):
         """Initialize a Task."""
         self._params = params
+        self.data = self.load_workflow_data()
```
Comment (chmeliik) [inline atomic_reactor/tasks/common.py:96]: Possibly a pro-tip for the future: fixing unit tests might have been less painful if this was a cached property. Currently, you need to mock context data just to initialize any Task. With a property, you'd only have to mock in the tests that actually use the data, which would probably be more or less the same tests that already do the mocking.

(No need to change it now)

File: atomic_reactor/tasks/common.py [hunk 2]
```diff
@@ -107,3 +112,31 @@ def load_workflow_data(self) -> inner.ImageBuildWorkflowData:
...
         return config.Configuration(self._params.config_file)
+
+    def throw_canceled_build_exception(self):
+        self.data.build_canceled = True
+        raise BuildCanceledException("Build was canceled")
+
+    @classmethod
+    def run(cls, init_build_dirs: Optional[bool] = False):
+        try:
+            cls.fs_watcher.start()
```
Comment (chmeliik) [inline atomic_reactor/tasks/common.py:123]: I don't think this should be a class method, fs_watcher is an instance attribute (and execute() and others as well)

File: atomic_reactor/tasks/common.py [hunk 3]
```diff
@@ -107,3 +112,31 @@ def load_workflow_data(self) -> inner.ImageBuildWorkflowData:
...
+
+            # non-plugin based tasks don't have the attribute
+            if hasattr(cls, 'keep_plugins_running') and cls.keep_plugins_running:
```
Comment (chmeliik) [inline atomic_reactor/tasks/common.py:126]: Maybe we could move this to Task level and rename? I can only think of silly names though, like `immortal` 😄

File: atomic_reactor/tasks/common.py (hunk 1 above)
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py:96]: hmm, would that not have detrimental effect when we want to actually update the property?

File: atomic_reactor/tasks/common.py (hunk 3 above)
Comment (chmeliik) [reply atomic_reactor/tasks/common.py]: Well we're using it here :smile: 

I imagine that if we reworked the exit task to not be plugin based, we would still want to ignore SIGTERM in it

File: atomic_reactor/tasks/common.py (hunk 1 above)
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py]: I always welcome pro-tips, updated.

File: atomic_reactor/tasks/common.py (hunk 3 above)
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py]: so what about `keep_task_running` or `cancellable_task`

Comment (chmeliik) [reply atomic_reactor/tasks/common.py]: `ignore_sigterm` would be a good name if it's just for ignoring SIGTERM

Come to think of it, maybe it *should* be just for that, and `keep_plugins_running` should be a separate thing that decides only if plugin errors will be ignored

Comment (mkosiarc) [reply atomic_reactor/tasks/common.py]: added `ignore_sigterm` class variable, since it should do different thing than `keep_plugins_running`

File: atomic_reactor/tasks/common.py [hunk 4]
```diff
@@ -87,9 +91,12 @@ def _drop_known_unset_args(cls, args: dict) -> dict:
 class Task(abc.ABC, Generic[ParamsT]):
     """Task; the main execution unit in atomic-reactor."""
 
+    ignore_sigterm: ClassVar[bool] = False
+
     def __init__(self, params: ParamsT):
         """Initialize a Task."""
         self._params = params
+        self.fs_watcher = FSWatcher()
```
Comment (mkosiarc) [inline atomic_reactor/tasks/common.py:99]: Hm, maybe not all tasks should have a FSWatcher() because the data from it are used only in store_metadata plugin to export it in annotations.

Comment (mkosiarc) [reply atomic_reactor/tasks/common.py:99]: @chmeliik @tkdchen what do you think?

Comment (chmeliik) [reply atomic_reactor/tasks/common.py:99]: Hmm, not gonna lie, I don't really know what `FSWatcher` is good for :smile: 

I guess it could be useful if we wanted to figure out how much disk space our builds are actually using, and for that, we would want to keep track of the usage across all tasks. Which would mean saving the results in context data somehow, and that sounds kinda complicated, so maybe it should be a separate ticket.

There might also be other options - e.g. just _log_ the usage after each task, which would make it easier to query in Splunk anyway. Or get rid of it entirely and rely on some OpenShift monitoring mechanism to get usage stats instead.

TL;DR, I think FSWatcher doesn't really need to do anything sensible at the moment and we should figure it out later

**Review by hjmodi — APPROVED** [review]

**Review by chmeliik — COMMENTED** [review]
LGTM, just a minor nitpick for a log message

And another fairly important change here is saving context data in every Task, maybe that should be in the commit message as well

File: atomic_reactor/tasks/common.py [hunk 5]
```diff
@@ -107,3 +112,26 @@ def load_workflow_data(self) -> inner.ImageBuildWorkflowData:
...
+
+        finally:
+            signal.signal(signal.SIGTERM, signal.SIG_DFL)
+            # For whatever the reason a task fails, always write the workflow
+            # data into the data file.
+            self.data.save(self.get_context_dir())
+
+    @cached_property
+    def data(self):
+        return self.load_workflow_data()
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py:137]: well, it is already done, so..:D But yes, this needed to be done so the context could be saved at the last level, see line `self.data.save(self.get_context_dir())`

File: atomic_reactor/tasks/common.py [hunk 6]
```diff
@@ -107,3 +112,26 @@ def load_workflow_data(self) -> inner.ImageBuildWorkflowData:
 
     def load_config(self) -> config.Configuration:
         return config.Configuration(self._params.config_file)
+
+    def throw_task_canceled_exception(self, *args, **kwargs):
+        self.data.task_canceled = True
+        raise TaskCanceledException("Task was canceled")
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py:118]: It means a-r task, which could be any task that inherits from class Task. that is why keeping the name `BuildCanceledException` does not seem right to me.

File: atomic_reactor/plugins/add_filesystem.py [hunk 7]
```diff
@@ -286,8 +286,8 @@ def run_image_task(self, image_build_conf):
         task = TaskWatcher(self.session, task_id, self.poll_interval)
         try:
             task.wait()
-        except BuildCanceledException:
-            self.log.info("Build was canceled, canceling task %s", task_id)
+        except TaskCanceledException:
+            self.log.info("Task was canceled, canceling task %s", task_id)
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/plugins/add_filesystem.py:290]: What about "Prebuild task was cancelled, cancelling brew task"? I am okay with "Build was cancelled, cancelling brew task" as well.

File: atomic_reactor/tasks/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/tasks/common.py:137]: Maybe we can just get rid of the `load_workflow_data()` method to avoid having one "wrong" way to load data? (@tkdchen is that what you meant?)

File: atomic_reactor/plugins/add_filesystem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/plugins/add_filesystem.py:290]: "Prebuild task" works for me :+1:

File: atomic_reactor/tasks/common.py [hunk 8]
```diff
@@ -107,3 +112,26 @@ def load_workflow_data(self) -> inner.ImageBuildWorkflowData:
...
+
+    def run(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/tasks/common.py:120]: Hmm, maybe `init_build_dirs` never should have been a parameter in the first place, but that's probably not worth fixing at the moment

I'm fine with *args **kwargs

File: atomic_reactor/plugins/add_filesystem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/add_filesystem.py:290]: tekton task was canceled, canceling koji task

File: atomic_reactor/tasks/common.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/tasks/common.py:118]: 'tekton task was cancelled'

[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py]: updated

File: atomic_reactor/plugins/add_filesystem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/plugins/add_filesystem.py]: updated

File: atomic_reactor/tasks/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkdchen) [reply atomic_reactor/tasks/common.py:137]: I thought if we add another another layer (a cached property) to wrap `load_workflow_data`, why not just convert `load_workflow_data` to a cached property (whatever the name is).

[the remark this answers is not in the document]
Comment (mkosiarc) [reply atomic_reactor/tasks/common.py]: updated to have only one cached property handling this

File: tests/tasks/test_common.py [hunk 9]
```diff
@@ -169,10 +168,47 @@ def test_load_workflow_data(self, tmp_path):
...
+
+    def workflow_data(self, params):
         expect_data = inner.ImageBuildWorkflowData()
         (flexmock(inner.ImageBuildWorkflowData)
          .should_receive("load_from_dir")
          .with_args(dirs.ContextDir)
          .and_return(expect_data))
 
-        assert self.ConcreteTask(params).load_workflow_data() == expect_data
+        assert self.ConcreteTask(params).workflow_data() == expect_data
```
Comment (chmeliik) [inline tests/tasks/test_common.py:182]: If this is meant to be a unit test, I think it needs to start with `test_`

**Review by chmeliik — APPROVED** [review]
LGTM :+1:
